pkg/mcs/scheduling/server/meta/watcher.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package meta
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/core"
+	"github.com/tikv/pd/pkg/storage/endpoint"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/mvcc/mvccpb"
+	"go.uber.org/zap"
+)
+
+// Watcher is used to watch the PD API server for any meta changes.
+type Watcher struct {
+	wg        sync.WaitGroup
+	ctx       context.Context
+	cancel    context.CancelFunc
+	clusterID uint64
+	// storePathPrefix is the path of the store in etcd:
+	//  - Key: /pd/{cluster_id}/raft/s/
+	//  - Value: meta store proto.
+	storePathPrefix string
+
+	etcdClient   *clientv3.Client
+	basicCluster *core.BasicCluster
+	storeWatcher *etcdutil.LoopWatcher
+}
+
+// NewWatcher creates a new watcher to watch the meta change from PD API server.
+func NewWatcher(
+	ctx context.Context,
+	etcdClient *clientv3.Client,
+	clusterID uint64,
+	basicCluster *core.BasicCluster,
+) (*Watcher, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	w := &Watcher{
+		ctx:             ctx,
+		cancel:          cancel,
+		clusterID:       clusterID,
+		storePathPrefix: endpoint.StorePathPrefix(clusterID),
+		etcdClient:      etcdClient,
+		basicCluster:    basicCluster,
+	}
+	err := w.initializeStoreWatcher()
+	if err != nil {
+		return nil, err
+	}
+	return w, nil
+}
+
+func (w *Watcher) initializeStoreWatcher() error {
+	putFn := func(kv *mvccpb.KeyValue) error {
+		store := &metapb.Store{}
+		if err := proto.Unmarshal(kv.Value, store); err != nil {
+			log.Warn("failed to unmarshal store entry",
+				zap.String("event-kv-key", string(kv.Key)), zap.Error(err))
```

Comment:
Do we need to call `ExtractStoreIDFromPath` to get the store ID here?

Reply:
Do you mean to add it to the log?